builtin/providers/aws/resource_aws_sns_platform_application.go

```diff
@@ -0,0 +1,192 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsSnsPlatformApplicationAPNS() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsSnsPlatformApplicationAPNSCreate,
+		Read:   resourceAwsSnsPlatformApplicationAPNSRead,
+		Update: resourceAwsSnsPlatformApplicationAPNSUpdate,
+		Delete: resourceAwsSnsPlatformApplicationDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"type": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					if value != "SANDBOX" {
+						errors = append(errors, fmt.Errorf(
+							"%q has unsupported value %q", k, value))
+					}
+					return
+				},
+			},
+			"platform_credential": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"platform_principal": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceAwsSnsPlatformApplicationAPNSCreate(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	platform := "APNS"
+	if d.Get("type") != nil {
+		platform += "_" + d.Get("type").(string)
+	}
+
+	resp, err := snsconn.CreatePlatformApplication(&sns.CreatePlatformApplicationInput{
+		Platform: aws.String(platform),
+		Name:     aws.String(d.Get("name").(string)),
+		Attributes: aws.StringMap(map[string]string{
+			"PlatformCredential": d.Get("platform_credential").(string),
+			"PlatformPrincipal":  d.Get("platform_principal").(string),
+		}),
+	})
+
+	if err != nil {
```

Comment:
There are cases where terraform creates infra, someone deletes it from the UI and then this will fail. In this instance, you may want to check that if the err is the equivalent of a 404, then to log it, d.SetId("") and then return nil as per this style example https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_ecr_repository.go#L68